### PR TITLE
Add deprecated rule warnings

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Ref: #6409

> Is there anything in the PR that needs further explanation?

To notify deprecation and make the migration easier.
Such warnings are added to a linting result and reporting output.

For example of a report:

```console
$ cd scripts
$ ../bin/stylelint.js visual.css --config=visual-config.json

Deprecation Warning: The "string-quotes" rule is deprecated and will be removed in the next major release.
Deprecation Warning: The "unit-case" rule is deprecated and will be removed in the next major release.

visual.css
 1:9   ✖  Expected "url("foo.css")" to be ""foo.css""       import-notation
 1:13  ✖  Expected single quotes                            string-quotes
 4:9   ✖  Unexpected invalid hex color "#4f"                color-no-invalid-hex
 6:1   ✖  Expected ".foo.bar" to have no more than 1 class  selector-max-class
 7:10  ✖  Unexpected unit "px"                              unit-disallowed-list
 8:3   ⚠  Expected custom property name to be kebab-case    custom-property-pattern

6 problems (5 errors, 1 warning)
```

### CODE REVIEW

File: lib/lintPostcssResult.js
Comment (jeddy3) [inline lib/lintPostcssResult.js:145]:
```suggestion
		`The "${ruleName}" rule is deprecated.`,
```

Let's keep it succinct in the terminal to avoid a wall of text.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@ybiquitous I've made one minor suggestion, otherwise LGTM.

File: lib/lintPostcssResult.js
Comment (Mouvedia) [reply lib/lintPostcssResult.js:145]: What about having just one sentence with a list instead of multiple sentences?
e.g. `The "string-quotes", "unit-case" and "foo-bar" rules are deprecated and will be removed in the next major release.`

Comment (ybiquitous) [reply lib/lintPostcssResult.js]: Thanks for the advice. I've simplified warning messages via ff83ba86c. Here is a screenshot:

image

I keep multiple sentences because each message may have a `reference`, and it's hard to include reference(s) in one sentence. See the code below:

https://github.com/stylelint/stylelint/blob/ff83ba86caa9d5b3f9db47903fbb29ec1f6bb98d/lib/formatters/stringFormatter.js#L59-L61

Instead, I've avoided repeated "Deprecation Warning" messages in each line and appended a " - " prefix to each line.

What about this change?

File: types/stylelint/index.d.ts
Comment (ybiquitous) [inline types/stylelint/index.d.ts:326]: [note] for more accuracte type.

File: lib/lintPostcssResult.js
Comment (Mouvedia) [reply lib/lintPostcssResult.js]: Let's take stylelint config standard as the baseline.
Let's say a user resets about 19% of the deprecated rules; that clocks at around 12 lines out of potentially 64.
In that scénario I would prefer a counter and the usage of `…` or `etc.` for more than 3.
e.g.
```js
`${count} deprecation warnings for rules ${rules[0]},  ${rules[1]}, ${rules[2]}, etc.`
```

File: types/stylelint/index.d.ts
Comment (Mouvedia) [reply types/stylelint/index.d.ts:326]: [question]

Should it be optional instead?
e.g. `reference?: string;`

i.e. reference must be set when you are explicit
e.g.
```
				deprecations: [
					{
						text: 'Deprecated foo',
						reference: undefined,
					},
				],
```

File: lib/lintPostcssResult.js
Comment (jeddy3) [reply lib/lintPostcssResult.js]: > What about this change?

SGTM, thank you.

File: lib/formatters/stringFormatter.js
Comment (jeddy3) [inline lib/formatters/stringFormatter.js:66]:
```suggestion
	return ['', yellow('Deprecation warnings:'), ...lines, ''].join('\n');
```

Minor nit, otherwise LGTM to merge.

**Review by jeddy3 — CHANGES_REQUESTED** [review]

File: types/stylelint/index.d.ts
Comment (ybiquitous) [reply types/stylelint/index.d.ts]: Good point. Fixed via d6108a993.

File: lib/formatters/stringFormatter.js
Comment (ybiquitous) [reply lib/formatters/stringFormatter.js]: Nice. Fixed via bec7cdc7c.

File: lib/lintPostcssResult.js
Comment (ybiquitous) [reply lib/lintPostcssResult.js]: @Mouvedia I understand your concern. But, if `reference` is specified with a rule deprecation, how should we display it with `…` or `etc.`?　I believe we need to consider also plugin rules, not only the built-in ones.

**Comment by ybiquitous** [comment]
Maybe off-topic, but I found the following description in our developer guide:

https://github.com/stylelint/stylelint/blob/f129f3b992accbf37c7d532d0d020c62c4f3749a/docs/developer-guide/rules.md?plain=1#L278

Indeed, the `stylelintReference` property seems not to be used:

```console
$ git grep -w 'stylelintReference' lib/
lib/createPartialStylelintResult.js:32:                reference: deprecationMessage.stylelintReference,
```

Should we follow the guideline still? If doing that, this PR change would show a large amount of text for users.

File: lib/lintPostcssResult.js
Comment (Mouvedia) [reply lib/lintPostcssResult.js]: Alright, as long as we a ENV variable—or something similar—that hides those warnings it should be OK.
@ybiquitous Does the `--quiet` option hide these warnings?

**Comment by ollie-iterators** [comment]
When will this be getting merged?

**Comment by jeddy3** [comment]
> When will this be getting merged?

The original author usually merges once a pull request has one approval. So, it'll be merged whenever @ybiquitous has time. We're all volunteers so things happen when they happen.

> Also, once this PR gets merged, will it cause https://github.com/stylelint/stylelint/pull/6573 to be able to be merged so that stylelint can be updated to version 15.0.0?

Yes, that's likely. I'll update https://github.com/stylelint/stylelint/issues/6409 to bring things together.

File: lib/lintPostcssResult.js
Comment (ybiquitous) [reply lib/lintPostcssResult.js]: > Does the `--quiet` option hide these warnings?

I'm afraid the answer is **no** because these warnings are not related to a *warning* severity produced by lint rules. We currently don't have a way to suppress this kind of warning, except for using a custom formatter.

**Comment by ybiquitous** [comment]
Because we will deprecate many rules in v15.0.0, some may see many deprecation warnings and feel annoyed. But I believe showing warnings should be better than showing nothing.

Also, if people update `stylelint` and `stylelint-config-{recommended,standard}` at the same time, they will not see the warnings, since the configs will remove deprecation rules.

We can consider adding an option to turn off deprecation warnings if people need it.

If there is no other feedback, I will merge this PR within a few days.

**Comment by jeddy3** [comment]
> Also, if people update `stylelint` and `stylelint-config-{recommended,standard}` at the same time, they will not see the warnings, since the configs will remove deprecation rules.

Yep, exactly. 

> If there is no other feedback, I will merge this PR within a few days.

Merging SGTM.

**Comment by ybiquitous** [comment]
> I will for the 15.1.0 version.

Let's open an issue when it becomes necessary.

File: lib/lintPostcssResult.js
Comment (Mouvedia) [inline lib/lintPostcssResult.js:144]: @ybiquitous

This PR was merged before #7550.
Should we use `process.emitWarning` instead?
i.e. `--no-deprecation` doesn't suppress these warnings

Comment (ybiquitous) [reply lib/lintPostcssResult.js:144]: Good point. We must consider backward compatibility if switching to `process.emitWarning` would make sense.

For example, we may need to update the following document:
https://github.com/stylelint/stylelint/blob/1ba1a0aeafaa124a4987a0135d511111cec3af0a/docs/developer-guide/rules.md?plain=1#L296-L301

Can you open a new issue for the discussion?

Comment (Mouvedia) [reply lib/lintPostcssResult.js:144]: I considered this not important since the deprecation was added 2 majors ago.
But [this](https://deps.dev/npm/stylelint/16.7.0/versions) made me realize that a lot dependents are still stuck on old majors.
e.g. WordPress/gutenberg#48142

> Can you open a new issue for the discussion?

OK.

**Pull request merged**
